Save for Later: update reader menu view model

ctarda opened the pull request:
Implements #9229 

This PR adds a section to the Reader headers to present the Saved for Later / Bookmarked items.

![bookmarks](https://user-images.githubusercontent.com/2722505/39510681-1cab06e6-4e1d-11e8-855b-96640d1cf485.jpg)

In the spirit of increasing testability and modularity, I have taken the opportunity to extract some logic out of `ReaderMenuViewController`, also with the intention of making the current implementation a bit more open for modification.

Caveats:
- In order to present the menu item, I have added a mock subclass of `ReaderAbstractTopic` called `ReaderSaveForLaterTopic`. This class should be a proper core data entity. and will have to be updated when the core data model is.
- Also, there is some hardcoded logic in `ReaderMenuViewModel` hidden behind a feature flag, that is marked as "to be removed" after the new topic is provided by the core data model.
- I am terrible at naming stuff. Also, @malinajirka has suggested that we should probably use "Bookmark/ed" instead of "Saved for Later", which I think makes sense. But in the meantime, here we are! 😛 

On the bright side... there are some unit tests!

To test:
1. Checkout the branch
2. Run it as it is, the extra menu item should be visible in `localDeveloper` mode.

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderMenuItemCreator.swift:
```diff
@@ -0,0 +1,5 @@
+/// Interface abstracting the entities that create menu items for the different topics presented in Reader
+protocol ReaderMenuItemCreator {
```
Because each class is so short, how would you feel about putting all of the `ReaderMenuItemCreator` implementations into a single swift file? (this one, I guess)

ctarda replied:
I lean towards applying the single responsibility principle to files as well: one abstraction per file. I guess I tend to see short, granular files as better than longer files, slightly less cohesive.  I do not have very strong feelings about it though, so if you'd really prefer to move everything into a single file, I'm happy to do it 😃

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderMenuViewModel.swift:
```diff
@@ -234,53 +260,42 @@ enum ReaderDefaultMenuItemOrder: Int {
+        // To be removed as soon as the topic is provided by the coredata store. This is here just to prove visually that the view model can handle this topic
+        if FeatureFlag.saveForLater.enabled {
+            let topic = ReaderSaveForLaterTopic()
+            topic.title = "Bookmarked - mock"
```
Based on the first iteration of designs we now have (even though this will be removed soon), let's rename this to "Saved Posts" for now.

frosty commented on WordPress/Classes/ViewRelated/Reader/SavedForLaterMenuItemCreator.swift:
```diff
@@ -0,0 +1,19 @@
+        //TODO. Update with the proper icon
+        var item = ReaderMenuItem(title: topic.title,
+                                  type: .topic,
+                                  icon: Gridicon.iconOfType(.star),
```
To match our latest designs, let's switch this to the `bookmark` Gridicon.

ctarda commented:
Yes, the crash is expected (for now, until we switch to the updated core data model). Sorry, I forgot to mention it in the PR description 😊

ctarda commented:
I've updated the PR, with the new icon, and extracted the topic title to an extension on String. The discussion about coalescing all the files into one remains open 😛

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderStrings.swift:
```diff
@@ -0,0 +1,6 @@
+/// Encapsulates user facing strings in Reader
+extension String {
+    static var savedForLaterMenuTitle: String {
```
Sorry for being awkward again, but I'm not a huge fan of public extensions to classes like `String` here, to add a very specific property that's only used in one place in the app. 

I think I'd prefer a private enum in the view model (could be called `Strings`), with static constants for any strings we need. It's the model we follow in a lot of places in the app at the moment, and it keeps the constants much closer to where we actually need them. Would you be open to that?

ctarda replied:
I agree. This was meant to be just the starting point, and meant to grow with more strings as we go. I'll move it to the view model

ctarda replied:
I made it a struct though. I am not a huge fan of scattering calls to NSLocalizedString, which add a lot of noise. An enum forces "raw values for enum case must be a literal", and to be honest, adding a method to the enum just to wrap the raw value in a call to NSLocalizedString seems a bit too much.

ctarda commented:
Oh! I totally misunderstood the way you were suggesting using an enum. That's neat, I hadn't thought of that. I'll switch it to an enum and merge 😉
